snorkel/models/context.py

```diff
@@ -158,7 +159,193 @@ def _asdict(self):
+    position = Column(Integer, nullable=False)
+
+    __mapper_args__ = {
+        'polymorphic_identity': 'row',
+    }
+
+    __table_args__ = (
+        UniqueConstraint(document_id, table_id, position),
+    )
+
+    def __repr__(self):
+        return "Row(Doc: %s, Table: %s, Position: %s)" % (self.document.name, self.table.position, self.position)
+
+
+class Col(Context):
+    """A column Context in a Document."""
+    __tablename__ = 'col'
+    id = Column(Integer, ForeignKey('context.id'), primary_key=True)
+    document_id = Column(Integer, ForeignKey('document.id'))
+    table_id = Column(Integer, ForeignKey('table.id'))
+    document = relationship('Document', backref=backref('cols', cascade='all, delete-orphan'), foreign_keys=document_id)
+    table = relationship('Table', backref=backref('cols', cascade='all, delete-orphan'), foreign_keys=table_id)
+    position = Column(Integer, nullable=False)
+
+    __mapper_args__ = {
+        'polymorphic_identity': 'col',
+    }
+
+    __table_args__ = (
+        UniqueConstraint(document_id, table_id, position),
+    )
+
+    def __repr__(self):
+        return "Col(Doc: %s, Table: %s, Position: %s)" % (self.document.name, self.table.position, self.position)
+
+
+class Cell(Context):
+    """A cell Context in a Document."""
+    __tablename__ = 'cell'
+    id = Column(Integer, ForeignKey('context.id'), primary_key=True)
+    document_id = Column(Integer, ForeignKey('document.id'))
+    table_id = Column(Integer, ForeignKey('table.id'))
+    row_id =  Column(Integer, ForeignKey('row.id'))
+    col_id =  Column(Integer, ForeignKey('col.id'))
+    document = relationship('Document', backref=backref('cells', cascade='all, delete-orphan'), foreign_keys=document_id)
+    table = relationship('Table', backref=backref('cells', cascade='all, delete-orphan'), foreign_keys=table_id)
+    row = relationship('Row', backref=backref('cells', cascade='all, delete-orphan'), foreign_keys=row_id)
+    col = relationship('Col', backref=backref('cells', cascade='all, delete-orphan'), foreign_keys=col_id)
+    row_num = Column(Integer)
+    col_num = Column(Integer)
+    text = Column(Text, nullable=False)
+    html_tag = Column(Text)
+    if snorkel_postgres:
+        html_attrs = Column(postgresql.ARRAY(String))
+        html_anc_tags = Column(postgresql.ARRAY(String))
+        html_anc_attrs = Column(postgresql.ARRAY(String))
+    else:
+        html_attrs = Column(PickleType)
+        html_anc_tags = Column(PickleType)
+        html_anc_attrs = Column(PickleType)
+
+    __mapper_args__ = {
+        'polymorphic_identity': 'cell',
+    }
+
+    __table_args__ = (
+        UniqueConstraint(document_id, table_id, row_id, col_id),
+    )
+
+    def __repr__(self):
+        return ("Cell(Doc: %s, Table: %s, Row: %s, Col: %s)" % 
+            (self.document.name, self.table.position, self.row.position, self.col.position))
```

Comment:
Move back to tree structure- `Cell`s are direct children of `Table`, and have `row` and `col` attributes.  Then just add helper methods e.g. `get_row_siblings` and `get_col_siblings` for convenience